File: control/surfaces.lua
local deepmerge = require("helpers/deepmerge")
local identifiers = require("identifiers")

---@param settings MapGenSettings
local function preventAutoplace(settings)
    return deepmerge(settings or {}, {
        property_expression_names = {
            ['entity:warp-console:probability'] = '-inf',
            ['tile:empty-space:probability'] = '-inf',
            ['tile:' .. identifiers.warpTile .. ':probability'] = "-inf"
        }
    })
end

local surfaces = {
    nauvis = {
        default = preventAutoplace(prototypes.map_gen_preset["default"].basic_settings)
    }
}

---@param planet string
---@param type string
---@return MapGenSettings
local function getSurface(planet, type)
    return deepmerge(surfaces.nauvis.default, {
        seed = math.random(0, (2 ^ 32) - 1)
    })
end

---@param player LuaPlayer
---@param surface LuaSurface
---@param force boolean | nil
local function teleportToSurface(player, surface, force)
    if not force and player.surface ~= surface then
        local pos = surface.find_non_colliding_position("character", player.position, 128, 1, true)
        if pos then
            player.teleport(pos, surface)
            return
        end
    end

    local pos = surface.find_non_colliding_position("character", {0, 1}, 128, 1, true)
    if pos then
        player.teleport(pos, surface)
        return
    end

    player.print("warp-teleport.no-position")
end

return { getSurface = getSurface, teleportToSurface = teleportToSurface }

File: control/surfaces.ts
import { setSurfaceClear, SurfaceClearAction } from "control/state";
import { LuaPlayer, LuaSurface } from "factorio:runtime";

export function createSurfaceFor(planet: string, action: SurfaceClearAction) {
    const surface = createNewSurface(planet);

    surface.map_gen_settings = {
        ...surface.map_gen_settings,
        seed: math.random() * 4294967295
    };
    
    surface.clear();

    setSurfaceClear(surface.index, action);
}

export function createNewSurface(planet: string) {
    return getInternalPlanet(planet).create_surface();
}

function getInternalPlanet(planet: string) {    
    const existing = game.planets[`${planet}-a`].surface;
    if (existing) {
        return game.planets[`${planet}-b`];
    }
    return game.planets[`${planet}-a`];
}

export function teleportToSurface(player: LuaPlayer, surface: LuaSurface, force?: boolean) {
    if (!force && player.surface != surface) {
        const pos = surface.find_non_colliding_position("character", player.position, 128, 1, true)
        if (pos) {
            player.teleport(pos, surface);
            return;
        }
    }

    const pos = surface.find_non_colliding_position("character", [0, 1], 128, 1, true)
    if (pos) {
        player.teleport(pos, surface);
        return;
    }

    player.print("warp-error.no-position");
}
